Dynamic registration

aviars opened the pull request:
I have added a the section at the bottom called "Communicating POET JWTs in an OAuth 2.0 Dynamic Registration Scenario".

I have added key to the POET payload `poet_public_key_uri` that can be used to allow the public key to be hosted on a subdomain of the `iss`. If this is omitted the default location is still `[iss]/.well-known/poet.key`.  I also changes the `iss` to a URI instead of just a hostname.

Requesting view from @jmandel  and @jricher

jmandel commented on README.md:
```diff
@@ -25,7 +25,8 @@ Example  JWT
 The  example signed JWT (JWS) contains infromation about the _Cardiac Risk App_ OAuth2 application and is signed by _nate-trust.org_.
-The JWS is signed with a private key using the `RS256 Algorithm`.  If an x509 certificate is used for signing, then the corresponding public certificate shall be hosted at `https://nate-trust.org/.welknown/poet.pem`.  If a bare key iss used, the corresponding public key shall be hosted at `https://nate-trust.org/.welknown/poet.jwks`.  
+The JWS is signed with a private key using the `RS256 Algorithm`.  The corresponding public key, whether an X509 certificate or a JWKS, shall be stored at the stored at the URI found in the `poet_public_key_uri` in the payload.  The 
+`poet_public_key_uri` must be the same domain or a subdomain as the domain found in `iss`.  For example, if the `iss` is `https://nate-trust.org`, then acceptable value for the `poet_public_key_uri` include `https://nate-trust.org/.well-known/poet.pem`.  `https://subdomain.nate-trust.org/.well-known/poet.pem`, `https://nate-trust.org/.well-known/poet.jwks` and `https://another-subdomain.nate-trus ...
```
I think this leaves a lot of room for surprising behavior with subdomains, etc. Would prefer to have a single rule the public key is published at a specified URL (or two, if we need pem + jwks support)

jricher replied:
Agreed -- this construction pattern sounds like it's got a lot of potential for abuse. Similar schemes have caused problems with redirect_uri and I'd be hesitant to take this approach.

jricher replied:
Why can't we trust discovery on the `iss` field here to give us the key? Carrying the key URI with the payload is asking for trouble as an attacker could potentially substitute their own. Let's say that I get "yournamehere.google.com", with this scheme I could put my own key and claim to be from google's `iss`.

jmandel replied:
Yes, I'm saying what @jricher is saying. The "discovery on the iss" still needs to be specified, but I think it's something like:

1. Try resolving `[iss]/.well-known/poet.jwks`
2. Else, try resolving `[iss]/.well-known/poet.pem`

jmandel commented on README.md:
```diff
@@ -94,18 +95,67 @@ List of All Fields in the Payload
+                    "software_id": "blood-pressure-app",
+                    "redirect_uri": ["https://apps-dstu2.smarthealthit.org/bp-centiles/redirect"],
+                    ...
+                    "poet_endorsements": ["JWT3", "JWT4"]
```
would make this `poet_endorsement` (singular noun) to be consistent with `redirect_uri` and other fields.

jmandel commented on README.md:
```diff
@@ -1,22 +1,22 @@
-3.	An OAuth2 Server (OAuth2 Provider) may obtain a list of JWTs (i.e. a manifest) from a central source such as URL. Optionally, when an application developer is registering an application in the OAuth2 server, he or she may optionally include one or more JWT to the application registration.  The endorsement badge information will be displayed to the developer upon registration.
+1.	A developer registers his or her application with the Endorsing Body (EB). Information collected includes many of the same elements used in an oAuth client application registration.  Items include the name, host name, and redirect URLs of the application.
+2.	When approved by the EB, these values become memorialized and are embedded into a signed JWT, containing a software statement, that is given to the developer.  The EB's application registry contains one JSON document per certified application.
+3.	An OAuth2 Server (OAuth2 Provider) may obtain a list of JWTs (i.e. a manifest) from a central source such as URL. Optionally, when an application developer is registering an application in the OAuth2 server, he or she may optionally include one or more JWT to the application registration as a software statement (?).  The endorsement badge information will be displayed to the developer upon reg ...
```
Should strip the `as a software statement (?)` here and replace it with a description of the proposed protocol (a new property in the payload, with an array containing a list of URLs).

jricher commented on README.md:
```diff
@@ -54,6 +55,7 @@ Example Payload
     "scope" : "openid profile patient/*.read",
     "token_endpoint_auth_method" : "none",
     "grant_types" : [ "authorization_code" ]
+    "poet_public_key_uri": "https://nate-trust.org/.well-known/poet.pem"
```
Same comment as above -- let this be discoverable from the issuer.

jricher commented on README.md:
```diff
@@ -94,18 +95,67 @@ List of All Fields in the Payload
+one or more  an OAuth2 client application.  OAuth2 Providers may obtain 
+URLs from trusted parties or from developers.  The format of the JSON document 
+is based on, RFC 7591, OAuth 2.0 Dynamic Registration. This profile defines 
+one optional, additional field  `poet_jwt_endorsements` that contains an array of POET JWT endorsements.
```
I don't fully understand this manifest file concept, especially if it's hosting more than one client's information. If I include the manifest URL in my registration, which client entry does it refer to? What happens if the provider changes the information at the URL? Does the client send these fields as well, and do they have to match up?

This feels mostly like "software statement by reference". What's the burden in including the endorsement JWTs with the client itself? We had some requests for doing it by reference in RFC7591 but decided to do it only by value as it's more predictable and straight forward to process.

jmandel replied:
I think we could write a spec to include endorsement JWTs directly in the client (strings with JWTs, rather than strings with URLs). Especially if we can describe a nice revocation protocol whereby any endorsement issuer hosts contents at a derivable URL (like `[iss]/.well-known/poet/revocations/[software_id].json`, or something similarly specific)...

jmandel replied:
(That said, there are other use cases in the ecosystem where a manifest file is helpful -- e.g. for one EHR's app store to pull in app listings from a common source, etc. That's not in POET scope though, IMHO.)

aviars commented:
I agree the subdomain could be abused.  It has been removed.  I have changed `poet_endorsements` to the singular `poet_endorsement` and pushed to this branch.

As for the application manifest file, I'm still unsure. @jricher brings up a good point. What happens if that information has changed in the file? I think the expectation is that if the information is misaligned in some way, then its invalid.  Perhaps the JSON file format is out of scope for POET, but should it be defined somewhere? I hear you @jmandel that there are many healthcare use cases for a file full of application information in a standardized format.  The main objective here is to streamline creating application registries, endorsements, and dynamic application registration.  Having this information at a uri in a predictable way is a good idea.

aviars commented:
I didn't see your recent replies before my reply.  Maybe it doesn't make sense for dynamic registration, but I think for key items we cant to use the field names from 7591 to the extent possible.  @jmandel I think what your saying is that the JSON object itself needs defined?

aviars commented:
It should be noted, that not all health care applications function with OAuth.  In fact, many do not.  `redirect_uri` makes no sense in those contexts. Knowing however, at a minimum, that the `client_name`, `client_uri`, and `logo_uri` are consistent would be helpful.

jmandel commented:
In our discussion (and in the README) let's separate out:

1. The way an app can include a list of endorsements JWTs in a dyn-reg request. This should be a simpler array of JWT strings.

2. The way an endorsing body (or anybody else) can host a manifest of endorsed apps. My take is that we're really not ready to standardize this yet, and don't need to for the current use case. But if we do standardize it, we should make sure that certain tasks (like "learn about the list of app endorsed by a particular group", and "check to see if this endorsement has been revoked") are straightforward.

jricher commented:
+1 to Josh's comment. The latter part, the manifest of endorsements, is similar to the client/RP discovery we tried in BB+REST but didn't get traction at that stage.

aviars commented:
Just pushed some changes to this branch:

- Added details for adding POET to a dyn-reg. 
- Removed manifest file format.

 @jmandel  I agree that trying to define that file format is premature, but yes something of the of the sort is needed just to setup a demo. @jricher Can you point to the BB+REST?

aviars commented:
Should the payload have a `sub`?  Should that be the same as the `client_name`?

aviars commented:
Is `initiate_login_uri`defined somewhere? is this needed?

aviars commented:
Pinging @jmandel and @jricher for review.  I think we are ready to close?

jmandel commented on README.md:
```diff
@@ -1,22 +1,22 @@
-3.	An OAuth2 Server (OAuth2 Provider) may obtain a list of JWTs (i.e. a manifest) from a central source such as URL. Optionally, when an application developer is registering an application in the OAuth2 server, he or she may optionally include one or more JWT to the application registration.  The endorsement badge information will be displayed to the developer upon registration.
+1.	A developer registers his or her application with the Endorsing Body (EB). Information collected includes many of the same elements used in an oAuth client application registration.  Items include the name, host name, and redirect URLs of the application.
+2.	When approved by the EB, these values become memorialized and are embedded into a signed JWT, containing a software statement, that is given to the developer.  The EB's application registry contains one JSON document per certified application.
+3.	An OAuth2 Server (OAuth2 Provider) may obtain a list of JWTs (i.e. a manifest) from a central source such as URL. Optionally, when an application developer is registering an application in the OAuth2 server, he or she may optionally include one or more JWT to the application registration as a POET Endorsement JWT.  The endorsement badge information will be displayed to the developer upon regis ...
```
I'd replace this entire bullet with:

```
When a client registers with a service providing via OAuth 2.0 Dynamic Registration,
the client may provide a `poet_endorsement_jwt` parameter as part of the registration
request body. This value is an array of POET endorsements, each represented as a
JWT string (see example 2 below for details).
```

That is, I would:
 * eliminate confusing language about a "manifest" which doesn't seem relevant here
 * eliminate information about "displayed to the developer" since dyn-reg is automated (API-driven)
 * rename the field to `poet_endorsement_jwt` (the other way seems backwards to me, inserting information about format in the middle of the name)

jmandel commented on README.md:
```diff
@@ -94,18 +93,30 @@ List of All Fields in the Payload
+                    "software_id": "cadiac-risk-app",
+                    "redirect_uri": ["https://apps-dstu2.smarthealthit.org/cardiac-risk/redirect"],
+                    ...
+                    "poet_jwt_endorsement": ["JWT1", "JWT2"]
```
Would rename to `poet_endorsement_jwt`

jmandel commented:
I've proposed a few more changes. Overall I'd recommend refactoring this page into:

 * For app developers
 * For endorsers

and separate out the steps that each group would use.
